csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -19,6 +19,182 @@
+      auto root_ids = getInputsInTargetDomain(id, c_root_domain);
+      for (auto root_id : root_ids) {
+        c_reshaped_ids.insert(root_id);
+      }
+    }
+  }
+
+  for (auto id : c_reshaped_ids) {
+    if (auto p_id = c2p.find(id); p_id != c2p.end()) {
+      p_reshaped_ids.insert(p_id->second);
+    }
+  }
+  return std::make_pair(p_reshaped_ids, c_reshaped_ids);
+}
+
+bool isSplitDivisible(IterDomain* id, Split* split) {
+  if (!split->factor()->isConstInt()) {
+    return false;
+  }
+  auto split_factor = split->factor()->evaluate().as<int64_t>();
+  if (split_factor == 1) {
+    return true;
+  }
+
+  if (!id->extent()->isConstInt()) {
+    return false;
+  }
+  auto id_extent = id->extent()->evaluate().as<int64_t>();
+  return id_extent % split_factor == 0;
+}
+
+void splitLike(TensorView* tv, int64_t axis, Split* ref_split) {
+  NVF_ERROR(
+      isSplitDivisible(tv->axis(axis), ref_split),
+      "Require the sharded ID to be divisible by the split factor. Got: ",
+      tv->axis(axis),
+      " and split factor: ",
+      ref_split->factor());
+  tv->outer_split(axis, ref_split->factor());
+}
+
+// Returns the number of DID axis on reshaped ids that were propagated to the
+// consumer.
+void shardViewOp(ViewOp* view_op, int64_t& did_pos) {
+  // This implementation asserts that only one sharding is applied on the
+  // reshaped ids. Inner split is not supported. The cases are:
+  // 1. Split reshape: [h] -> [a, h/a]. Sharding on h is applied to a in
+  // consumer.
+  // 2. Merge reshape: [a, h/a] -> [h]. Sharding on a is applied to h in
+  // consumer.
+  // 3. Multiple splits or merge reshapes: [x, y, z] -> [xyz]. Sharding on x and
+  // xyz. Similarly for the corresponding split reshape.
+  // 4. Independent splits or merge reshapes: [w, x, y, z] -> [wx, yz]. Sharding
+  // is on w and y. In the consumer, it is applied to wx and yz.
+  // An improvement is to support mult-levels of sharding (not a real case yet)
+  // if they are all outer splits. For example: For the reshape [h] -> [a, h/a]
+  // where the h is sharded twice: [h] -> [cp, h/cp] -> [cp, tp, h/(cp*tp)]
+
+  // A more general approach maybe to "undo" the reshape (reverse transforms
+  // from root to logical domain), followed by simplification of the consumer
+  // loop domain to move DID upwards.
+
+  TensorView* producer = view_op->in();
+  TensorView* consumer = view_op->out();
+
+  const std::unordered_map<IterDomain*, IterDomain*>& c2p =
+      PairwiseLogicalDomainMap(producer, consumer).mapConsumerToProducer();
+  const std::unordered_map<IterDomain*, IterDomain*>& p2c =
+      PairwiseLogicalDomainMap(producer, consumer).mapProducerToConsumer();
+  auto [p_logical_reshaped_ids, c_root_reshaped_ids] =
+      getReshapedIds(view_op, c2p);
+
+  auto p_loop_domain = producer->getLoopDomain();
+
+  // Track number of DID axis on reshaped ids that were propagated to the
+  // consumer. These will not be included in TransformPropagator.
```

Comment:
Is the comment outdated? We no longer use TransformPropagator. Correct? 

And there are several other mentions in this file.